controllers/idler/idler_controller.go

```diff
@@ -133,9 +134,20 @@ func (r *Reconciler) ensureIdling(ctx context.Context, idler *toolchainv1alpha1.
 	newStatusPods := make([]toolchainv1alpha1.Pod, 0, 10)
 	for _, pod := range podList.Items {
 		pod := pod // TODO We won't need it after upgrading to go 1.22: https://go.dev/blog/loopvar-preview
-		logger := log.FromContext(ctx)
-		podLogger := logger.WithValues("pod_name", pod.Name, "pod_phase", pod.Status.Phase)
+		podLogger := log.FromContext(ctx).WithValues("pod_name", pod.Name, "pod_phase", pod.Status.Phase)
 		if trackedPod := findPodByName(idler, pod.Name); trackedPod != nil {
+			// check the restart count for the trackedPod
+			restartCount := getHighestRestartCount(pod.Status)
+			if restartCount > RestartThreshold {
+				podLogger.Info("Pod is restarting too often. Killing the pod", "restart_count", restartCount)
+				lctx := log.IntoContext(ctx, podLogger)
+				// Check if it belongs to a controller (Deployment, DeploymentConfig, etc) and scale it down to zero.
+				err := deletePodsAndCreateNotification(ctx, lctx, pod, r, idler)
```

Comment:
could you please clarify what is the point of using two types of context? Why not using one?
Also, I would rename the context to make it clear that it's related to the pod
```suggestion
				podCtx := log.IntoContext(ctx, podLogger)
				// Check if it belongs to a controller (Deployment, DeploymentConfig, etc) and scale it down to zero.
				err := deletePodsAndCreateNotification(podCtx, pod, r, idler)
```

Reply:
I tried answering it here - https://github.com/codeready-toolchain/member-operator/pull/630#discussion_r1970555406
To add though, podCtx would have the pod_name and pod_phase which is useful to have in the logs when scalling the controller and pods down. instead of passing it in the logs everytime. 
Below is the snippet of logs with normal ctx and podCtx. I'd keep the pod context - its helpful to see which pod is referenced in the log itself.
i'll rename the variable to podCtx though to make it clearer - thanks!
```
2025-02-26T13:52:38-05:00	INFO	Scaling controller to zero	{"pod_name": "restartCount-alex-stage-pod-fail", "pod_phase": "", "name": "restartCount-alex-stage-pod-fail"}
2025-02-26T13:52:38-05:00	INFO	Deleting pod without controller
2025-02-26T13:52:38-05:00	INFO	Pod deleted
```

Reply:
we could just use podCtx for the entire function, but it would be an overkill imo - for example: pod_name and pod_phase is not needed in the logs when creating a notification, it'll be good to have sure but i think we can get by without it too.

Reply:
I'm completely fine with using `podCtx`, which is also what I proposed, but I don't see any reason for passing and using the original `ctx` context - this is what I'm still missing in your explanation. 

> we could just use podCtx for the entire function, but it would be an overkill imo - for example: pod_name and pod_phase is not needed in the logs when creating a notification, it'll be good to have sure but i think we can get by without it too.

overkill is using two types of contexts if not really needed for any cancellation reasons or anything like that. 
This whole function `deletePodsAndCreateNotification` is executed in the context of the specific pod, so it's completely valid (and TBH also expected) to include the pod information in the logs. From your snippet, you can also see that the lines:
```
2025-02-26T13:52:38-05:00	INFO	Deleting pod without controller
2025-02-26T13:52:38-05:00	INFO	Pod deleted
```
are missing which pod it is referring to, which is more a bug than a feature. 
Providing the context values is the best practice we do everywhere else, just look at the controller context and the logger - it contains the request metadata of the object the reconcile was triggered for in the whole reconcile loop.